portfolio/src/main/client/src/util/html.ts

```diff
@@ -0,0 +1,153 @@
+  // combine items first so that arrays can be flattened
+  // without messing up indices
+  const combined: InterpolationValue[] = [fragments[0]];
+
+  for (let i = 0; i < items.length; i++) {
+    combined.push(items[i]);
+    combined.push(fragments[i + 1]);
+  }
+
+  const flattened = flatten(combined);
+
+  /**
+   * @param item The item to get the sentinel value for.
+   * @param idx The index of the item within the items array.
+   * @returns The sentinel value.
+   */
+  function getSentinel(
+    item: InterpolationValue,
+    idx: number): string {
+    if (typeof item === 'function') {
+      return `$$template_fn_${idx}`;
+    } else if (item instanceof Node || Array.isArray(item)) {
+      return `<slot data-template data-index="${idx}" />`;
+    } else if (item !== null && item !== undefined) {
+      return item.toString();
+    } else {
+      return '';
+    }
+  }
+
+  const markup = flattened.map(getSentinel).join('');
+
+  const template = document.createElement('template');
+
+  template.innerHTML = markup.trim();
+
+  for (const descendant of descendants(template.content)) {
+    if (!(descendant instanceof Element)) continue;
+
+    // use slots for substitution
+    if (descendant.tagName === 'SLOT') {
+      const slot = descendant as HTMLSlotElement;
+      if (slot.hasAttribute('data-template')) {
+        const index = parseInt(slot.getAttribute('data-index')!, 10);
+        const item = flattened[index];
+        // cannot be event handler b/c it uses a different sentinel
+        // cannot be null or undefined b/c those don't produce sentinels
+        slot.replaceWith(item as string | Node);
+      }
+    }
+
+    // use all @attributes as event handlers
+    for (const attr of descendant.getAttributeNames()) {
+      if (!attr.startsWith('@')) continue;
+      const event = attr.slice(1);
+
+      const sentinel = descendant.getAttribute(attr);
+      if (!sentinel) throw new Error(`Attribute ${attr} should have a function handler.`);
+
+      const match = /^\$\$template_fn_(\d+)$/.exec(sentinel);
+      if (!match) throw new Error(`Attribute ${attr} should have a function handler.`);
+
+      const index = parseInt(match[1], 10);
+      descendant.addEventListener(event, flattened[index] as EventListener);
+    }
+  }
+
+  return Array.from(template.content.children);
+}
+
+// eslint-disable-next-line valid-jsdoc
+/**
+ * Helper function to parse HTML into a DOM node. Meant to be used
+ * as a template literal tag:
+ * ```
+ * const myElem = htmlElement`<td>this is the html</td>`;
+ * ```
+ * @returns The HTML element.
+ */
+export function htmlElement<T extends Node>(
+  fragments: TemplateStringsArray,
+  ...items: InterpolationValue[]
+): T {
+  const result = htmlFragment(fragments, ...items);
+  if (result.length !== 1) {
+    throw new Error(
+      'This HTML should contain exactly one element. ' +
+      'For multiple elements, use htmls.');
+  }
+  return result[0] as T;
```

Comment:
Same as previous PR. Add an `result[0] instanceof T` assert here so you aren't using `foo as T`.

Reply:
Type parameters get stripped out at compile time, so you can't use them in `instanceof` tests which run at runtime.

Reply:
Ah, good point. Resolving.